App/StackExchange.DataExplorer/Controllers/AccountController.cs

```diff
@@ -475,6 +521,135 @@ private ActionResult FetchFromGoogle(string accessToken)
             }
         }
 
+        private bool TryGetStackAppsConfig(out string secret , out string clientId, out string path)
+        {
+            secret = AppSettings.StackAppsOAuthSecret;
+            clientId = AppSettings.StackAppsClientId;
+            path = "/user/oauth/stackapps";
+            return AppSettings.EnableStackAppsAuth;
+        }
+
+        [StackRoute("user/oauth/stackapps")]
+        public ActionResult StackAppsCallback(string code, string state, string error)
+        {
+            if (code.IsNullOrEmpty() || error == "access_denied")
+            {
+                return LoginError("(From StackApps) Access Denied");
+            }
+            string secret, clientId, path;
+            if (!TryGetStackAppsConfig(out secret, out clientId, out path))
+            {
+                return LoginError("StackApps Auth not enabled");
+            }
+
+            // Verify state
+            var oAuthState = JsonConvert.DeserializeObject<OAuthLoginState>(state);
+            var hash = (AppSettings.OAuthSessionSalt + ":" + oAuthState.ses).ToMD5Hash();
+            if (oAuthState.hash != hash)
+            {
+                return LoginError("Invalid verification hash");
+            }
+
+            var postForm = HttpUtility.ParseQueryString("");
+            postForm["code"] = code;
+            postForm["client_id"] = clientId;
+            postForm["client_secret"] = secret;
+            postForm["redirect_uri"] = BaseUrl + path;
+
+            for (var retry = 0; retry < GoogleAuthRetryAttempts; retry++)

```

Comment:
The naming of this const `GoogleAuthRetryAttempts` is somewhat confusing in this context. +1 for re-use but in 6 to 8 weeks you'll be so confused ...